src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
+        /// entry and update the first entry to:
+        /// 
+        ///     outputMap[foo_ndoe_mode_guid] = "foo21".
+        /// 
+        /// Renaming map are similar to output map. The difference is these
+        /// variables used internally and they may be renamed as well. In 
+        /// some sense, it is just subset of outputMap.
+        /// </summary>
+        /// <param name="nodes"></param>
+        /// <param name="inputMap"></param>
+        /// <param name="outputMap"></param>
+        /// <param name="renamingMap"></param>
+        private static void GetInputOutputMap(
+            IEnumerable<NodeModel> nodes, 
+            out Dictionary<string, string> inputMap,
+            out Dictionary<string, string> outputMap,
+            out Dictionary<string, string> renamingMap)
+        {
+            outputMap = new Dictionary<string, string>();
+            renamingMap = new Dictionary<string, string>();
+            var inputVariableSet = new HashSet<string>();
+
+            foreach (var node in nodes)
+            {
+                // Collects all inputs of selected nodes
+                foreach (var inport in node.InPorts)
+                {
+                    if (inport.Connectors.Count == 0)
+                        continue;
+
+                    var startPort = inport.Connectors[0].Start;
+                    var inputNode = startPort.Owner;
+                    var portIndex = startPort.Index; 
+                    var cbn = inputNode as CodeBlockNodeModel;
+
+                    if (nodes.Contains(inputNode))
+                    {
+                        // internal node
+                        if (cbn != null)
+                        {
+                            string inputVar = cbn.GetAstIdentifierForOutputIndex(portIndex).Value;
+                            string originalVar = cbn.GetRawAstIdentifierForOutputIndex(portIndex).Value;
+
+                            renamingMap[inputVar] = originalVar;
+                            var key = String.Format("{0}%{1}", originalVar, cbn.GUID);
+                            renamingMap[key] = inputVar;
+                        }
+                    }
+                    else // extern node, so they should be added to inputMap
+                    {
+                        string inputVar = inputNode.GetAstIdentifierForOutputIndex(portIndex).Value;
+                        inputVariableSet.Add(inputVar);
+                    }
+                }
+
+                // Collect all outputs of selected nodes
+                var thisCBN = node as CodeBlockNodeModel;
+                if (thisCBN != null)
+                {
+                    for (int i = 0; i < thisCBN.OutPorts.Count(); ++i)
+                    {
+                        var inputVar = thisCBN.GetAstIdentifierForOutputIndex(i).Value;
+                        var originalVar = thisCBN.GetRawAstIdentifierForOutputIndex(i).Value;
```

Comment:
Hi @ke-yu, can you help me understand this one (perhaps with a comment or a note here)? if my CBN's GUID is `1234`, and `inputVar = temp5678_0_1234`, `originalVar = temp5678`, what is the relationship between the two? Or in another word, how does a node outside of this CBN see/use these variables?